### ISSUE TITLE
Questions about widgets

### ISSUE THREAD

**Issue by marcinnformula** [django-import-export/django-import-export#1635]
Why DateWidget and DateTimeWidgets are ~~using input formats (settings.DATE_INPUT_FORMATS)~~ formatted string for output? Also, output formats should be dependent on output file type, because they produces wrong result where xls writer is getting string instead of datetime. I.e. for CSV it could format dt as string (localized or in format provided), and for xls it should passthrough (date)time objects. Moreover, widgets aren't configurable via `__init__`, which makes they usage very limited.

If there is consent for this type of changes, I can prepare several pull requests.

EDIT: I misinterpreted input formats usage - it is fine

**Comment by matthewhegarty** [comment]
Thanks for raising this.  We are working on v4 release now, so we would be open to changes to improve widget output.  We'd need to be careful to be consistent, so I would suggest proposing the changes first so that we can agree before you take on a lot of work.  

There is a similar change for [`CharWidget`](https://github.com/django-import-export/django-import-export/pull/1623) which has gone into v4, so configuring output via kwargs is something we can consider.

**Comment by marcinnformula** [comment]
> There is a similar change for [`CharWidget`]

Looks great.

> We'd need to be careful to be consistent, so I would suggest proposing the changes first

I understand.

I see that Char, Int, Float, Decimal, Boolean widgets are returning their native types from `render()`, but Date-related widgets returns strings. That's inconsistent. This may be expected for csv-like or screen outputs, but, for example, XLS requires more fancy value. Also note, that tablib works fine with datetime objects

I should not declare `ExceltDateTimeWidget` in my resource, because it won't fork for other formats. The conversion between native date/datetime values should be somewhere else, probably somewhere around tablib. So by default DateWidget/DateTimeWidgets should return native types (probably timezone aware until something like `use_naive` will be passed as argument). Then, something like tablib, should convert native types to proper format (-> widgets should not format dates).

In my case I must omit tablib comlpetely. It can't handle xls dates / datetimes properly https://github.com/jazzband/tablib/issues/210 But because Date and DateTime widgets are returing strings instead of native objects, I must also redeclare fields with my custom widgets which would not convert date/datetime objects into strings.

It would be nice to have consistent behaviour ootb.

**Comment by matthewhegarty** [comment]
Thanks for the comments.  v4 will be a good opportunity to refactor this, however there are a lot of implementations out there which rely on widgets, so we would have to be careful how we introduce any changes.  The codebase has evolved over a decade, and has had many changes from a number of contributors, hence why it is not completely consistent.

I mostly agree with what you are saying.  The present implementation uses widgets for the import / export conversions, where `clean()` translates to python value, and `render()` formats for export.  This is why the date is formatted in this function.  We would have to retain this behaviour, if you introduced tablib at this point, it would be another layer doing conversion, which would be unnecessary imo.

 tablib is involved further up the chain, so I think that widgets should remain separate from tablib.  Functions such as `export()` will know about file formats such as 'xlsx' and can therefore perhaps configure the widget with args to call `export()` to return a native type for Excel.

In summary I would favour consistent kwargs used to control output in the same manner as #1623.  This would be easy to implement, document and test without adversely affecting other implementations.

Related discussion: #1485

**Comment by marcinnformula** [comment]
> it would be another layer doing conversion, which would be unnecessary imo.

It is already there, and is perfectly fine. 

Django import-export has nothing to do with output format. It is a responsibility of tablib, which is doing (almost) all necessary conversions. But it can't handle date and datetimes correctly (apart from it's bug mentioned earlier), because it is getting strings from django-import-export as input data.

----

I wrote my custom exporter to xls to avoid tablib bugs/limitations. But when using default DateTime/Date widgets I'm getting strings from Resource. 

To avoid additional parsing from string to native values, and also additional formatting after converting native value to excel-compatible, I must declare custom widgets for my date and datetime fields. These custom widgets are workarounds for returing strings from Date/DateTime.render()

----

> I think that widgets should remain separate from tablib

Definitely. They should not have anything related to specific output file format.

> I would favour consistent kwargs used to control output in the same manner as https://github.com/django-import-export/django-import-export/pull/1623

I'm not sure about that.

Maybe Field or Widget can hold extra arguments for renderer, something like Django's Widget in `attrs`. They're meaningless for Django, but important for HTML.  So in your case `atts` introduced in Field's `Widget` can be optionally declared and used by renderers like tablib.

But this is a separate problem and we can discuss it somewhere else.

**Comment by matthewhegarty** [comment]
Thanks for clarifying.  You may well be describing an optimal implementation for this, but we are bound somewhat by the legacy codebase, and the existing implementations.  We don't get too many issues relating to Widgets, but that doesn't mean they can't be improved.  The codebase has evolved over time, so it could well be done differently if we were starting from scratch.

Over the next couple of major releases, hopefully we will be refactoring to improve extensibility, but we are a very small group, so it could take a long time, if it ever happens at all.

Just to be clear we would consider PRs along the lines I described to help modify the output of Widgets, but we would not be keen on larger scale refactoring, unless we were in complete agreement about what it would entail, so something to bear in mind before submitting any PRs.  

Thanks for your input.

**Comment by marcinnformula** [comment]
Thanks.

>  we would not be keen on larger scale refactoring

It's not my intention.

The worst thing is that the format of date/datetimes in the output files may change a little, as conversion will be delegated to tablib. I will have in mind backward compatibility.

**Comment by matthewhegarty** [comment]
I may have misunderstood but I would not be keen on delegating to tablib from widgets.  The `render()` method should be used to control the output, i.e. we'd want to stick to the existing definition of [widget interface](https://github.com/django-import-export/django-import-export/blob/09a08cd926271b406231f1bdfc182d86082323f8/import_export/widgets.py#L24).

If you do need this functionality, then you'd be better doing it in subclasses outside of the django-import-export implementation.

**Comment by marcinnformula** [comment]
> I may have misunderstood

Me too, probably. I will try to explain this.

>  I would not be keen on delegating to tablib from widgets.

You're already doing that for Char, Integer, Float, Decimal, Boolean widgets. 
And this is fine, because the final representation of these values is dependent on the output format. 
The output format is declared by taglib via dataset.xls, dataset.csv and so on, where dataset belongs to taglib.

But in some of your widgets you're converting values to strings, which leds to issues with some output formats.  I.e. xls output is broken because of rendering date and datetimes as strings. Tablib/xlwr has no chance to properly handle them.

For example - you want to have summable integeres in csv/xls, so you're rendering `int` type, right?

> If you do need this functionality, then you'd be better doing it in subclasses 

But what for? I have to built a workaround by such subclasses.
This is a potential fix for the issue we're discussing here:

```
  class FixedDateTimeWidget(DateTimeWidget):                                                                         
      # pylint: disable=unused-argument                                                                              
      def render(self, value, obj=None):                                                                             
          if not value:                                                                                              
              return None                                                                                            
          return value                                                                                               
                                                                                                                     
                                                                                                                     
  class FixedDateWidget(DateWidget):                                                                                 
      # pylint: disable=unused-argument                                                                              
      def render(self, value, obj=None):                                                                             
          if not value:                                                                                              
              return None                                                                                            
          try:                                                                                                       
              return value.date()                                                                                    
          # pylint: disable=broad-exception-caught
          except Exception:                                                                                          
              return value       
```

Both of them just avoids problematic `format_datetime` implemented in paretn's `render()`.
And their implementation is consistent with Boolean, Char, Decimal, Number, Float widgets.

----

I think that the original concept of rendering dates and datetimes as strings came from bugs in tablib. It has broken handling of dates when outputting xls. It is not declaring cells format to date/datetime, and instead of properly formatted date/datetime, the number is just displayed. It may have also issues with tz-aware datetimzes. 

Django import-export tries to be consistent here by converting date/datetimes to strings, but this is just a workaround. If not, I would ask then why Boolean/Char/Decimal/Number/Float widgets are not rendering strings in same way as Date/DatetTime widgets.

Note that Date/DateTime widgets have passthrough code in `clean` like this:
```
        if isinstance(value, date):
            return value
```
so they can handle date/datetimes read from xls files, interpreted by readers like xlrd as native types.

----

I know there are JsonWidgets and so on. Leave them alone for a while

**Comment by matthewhegarty** [comment]
OK thanks - I think that explains it for me.  

I think the basis of the issue is that in some contexts we need widget output to be either a string or native.  For example, some users want a formatted string (e.g. local date format in csv), but this doesn't work for xls / xlsx, because this format requires the native value.

Therefore that implies that we need to support both formats.  So I think your fix for `DateTimeWidget` is ok, but we'd need to support both native and string formats.  The change sounds similar to #1501, where we introduced formatting whilst retaining the native value.  So I guess it would be very easy to add similar logic to `DateTimeWidget`, along the lines of your fix.

So with that in mind, we could add `coerce_to_string` to `DateWidget` / `DateTimeWidget`, and then return the string version if it is set, otherwise return native.  Would that work for you?

btw I expect that returning native types for the other widgets has been fine because they get rendered correctly in string contexts as well as native.

**Comment by marcinnformula** [comment]
> So with that in mind, we could add `coerce_to_string` to `DateWidget` / `DateTimeWidget`, and then return the string version if it is set, otherwise return native. Would that work for you?

Yes, it would. 

I have trouble what default value should be used for `coerce_to_string` of DateTime/Date widgets. For best results it should default to `False`. For backward compatibility - to `True`.

You mentioned a new major version. Maybe you could break compatibility for the sake of better default date/datetime handling? Just thinking out loud.

**Comment by matthewhegarty** [comment]
> You mentioned a new major version. Maybe you could break compatibility for the sake of better default date/datetime handling?

Yes, this would be ok.  We would have to be consistent, and since `NumberWidget` default for `coerce_to_string` is `False`, we should do the same for `DateTimeWidget` / `DateWidget` imo.

**Comment by marcinnformula** [comment]
Thank you. Sounds excellent for me.

Would you like me to do the PR or would you rather code it yourself?
Just asking. Sometimes is easier/faster to DIY than talking about proposal/patch.
I'm not in a hurry - I've already worked around this problem. 
BTW: Thank you for taking the time to discuss this and for the itself. This is very useful.

**Comment by matthewhegarty** [comment]
Likewise thanks for your input, always good to raise issues and discuss how we can improve things.

Feel free to submit a PR (contributing guidelines are [here](https://github.com/django-import-export/django-import-export/blob/main/docs/contributing.rst), if you haven't already seen them).  I'm working on a few other issues with the library at the moment, so if you can submit the PR that would be great.  We're always on the look out for additional help, so if you wanted to pick up any of the other issues you'd be very welcome to do so.

**Comment by matthewhegarty** [comment]
> Would you like me to do the PR or would you rather code it yourself?

Feel free to submit this PR if you get a chance.  If you're not able to at the moment, no problem, please let me know.

**Comment by matthewhegarty** [comment]
Addressed in #1657
